src/object/srv_obj.c

```diff
@@ -1721,21 +1708,162 @@ obj_local_rw_internal(crt_rpc_t *rpc, struct obj_io_context *ioc,
 
 	rc = obj_rw_complete(rpc, ioc, ioh, rc, dth);
 	if (iods_dup != NULL)
-		daos_iod_recx_free(iods_dup, orw->orw_nr);
+		daos_iod_recx_free(iods_dup, iods_nr);
 	return unlikely(rc == -DER_ALREADY) ? 0 : rc;
 }
 
+/* Extract local iods/offs/csums by orw_oid.id_shard from @orw */
 static int
-obj_local_rw(crt_rpc_t *rpc, struct obj_io_context *ioc,
-	     daos_iod_t *split_iods, struct dcs_iod_csums *split_csums,
-	     uint64_t *split_offs, struct dtx_handle *dth)
+obj_get_iods_offs_by_oid(daos_unit_oid_t uoid, struct obj_iod_array *iod_array,
+			 struct daos_oclass_attr *oca, uint64_t dkey_hash,
+			 uint32_t layout_ver, daos_iod_t **iods,
+			 uint64_t **offs, struct dcs_iod_csums **csums, uint32_t *nr)
 {
-	int	rc;
-	int	count = 0;
+	struct obj_shard_iod	*siod;
+	uint32_t		local_tgt;
+	uint32_t		oiod_nr;
+	int			i;
+	int 			idx = 0;
+	int			rc = 0;
+
+	oiod_nr = iod_array->oia_oiod_nr;
+	D_ASSERT(oiod_nr > 0);
+	if (oiod_nr > 1 || *iods == NULL) {
+		D_ALLOC_ARRAY(*iods, oiod_nr);
+		if (*iods == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+
+		D_ALLOC_ARRAY(*offs, oiod_nr);
+		if (*offs == NULL)
+			D_GOTO(out, rc = -DER_NOMEM);
+
+		if (csums != NULL) {
+			D_ALLOC_ARRAY(*csums, oiod_nr);
+			if (*csums == NULL)
+				D_GOTO(out, rc = -DER_NOMEM);
+			for (i = 0; i < oiod_nr &&
+			     iod_array->oia_iods[i].iod_type == DAOS_IOD_SINGLE; i++) {
+				D_ALLOC_PTR((*csums)[i].ic_data);
+				if ((*csums)[i].ic_data == NULL)
+					D_GOTO(out, rc = -DER_NOMEM);
+			}
+		}
+	}
+
+	local_tgt = uoid.id_shard % obj_ec_tgt_nr(oca);
+	for (i = 0; i < oiod_nr; i++) {
+		daos_iod_t		*iod_parent;
+		struct dcs_iod_csums	*iod_pcsum = NULL;
+		struct obj_io_desc	*oiod;
+
+		iod_parent = &iod_array->oia_iods[i];
+		oiod = &iod_array->oia_oiods[i];
+		if (iod_parent->iod_type == DAOS_IOD_ARRAY) {
+			siod = obj_shard_iod_get(oiod, local_tgt);
+			if (siod == NULL)
+				continue;
+		}
+
+		memcpy(&(*iods)[idx], iod_parent, sizeof(daos_iod_t));
```

Comment:
original code seems need not do such memcpy for iod/csum on tgt server, it just decode normal iod/csum from RPC, TBH I'm not sure why you think change like this is better?

Reply:
Hmm, I think the original code also need this extraction the iods/csums from the parent RPC, see the bottom half of obj_ec_rw_req_split().  no?